File: src/main/kotlin/alg02deques/ArrayDeque.kt
package alg02deques

class ArrayDeque<T>(capacity: Int = 2) : Deque<T> {
    private object Empty

    private var array = Array<Any?>(capacity) { Empty }
    private var firstItemIndex = 0
    private var lastItemIndex = 0

    private var _size: Int = 0

    override val size: Int
        get() = this._size

    private val capacity: Int
        get() = this.array.size

    private fun assertIndex(index: Int) {
        if (index < 0 || index >= this.size) {
            throw IndexOutOfBoundsException("$index is out of bounds")
        }
    }

    private fun normalizeIndex(index: Int): Int {
        this.assertIndex(index)
        /*
        As first and last item indexes are wrapped around at the end of the array
        we need to normalize indexes that come in range 0..size-1
        to be in range firstItemIndex..lastItemIndex-1
         */
        return (this.firstItemIndex + index).rem(this.capacity)
    }

    override operator fun get(index: Int): T {
        @Suppress("UNCHECKED_CAST")
        return this.array[normalizeIndex(index)] as T
    }

    override operator fun set(index: Int, item: T) {
        this.array[normalizeIndex(index)] = item
    }

    private fun resize(capacity: Int) {
        val array = Array<Any?>(capacity) { Empty }
        for (index in 0 until this._size) {
            array[index] = this[index]
        }
        this.array = array
        this.firstItemIndex = 0
        this.lastItemIndex = this._size
    }

    private fun growIfRequired() {
        if (this._size == this.capacity) {
            this.resize(this.capacity * 2)
        }
    }

    private fun shrinkIfRequired() {
        /*
        If array is quarter full it is shrunk to be half-size.
        It is resized at quarter capacity to avoid "thrashing" (constant resizing)
        when working with half-full array and doing add-remove operations.
        */
        if (this._size > 0 && this._size == this.capacity / 4) {
            this.resize(this.capacity / 2)
        }
    }

    override fun addFirst(item: T) {
        this.growIfRequired()
        this.firstItemIndex--
        if (this.firstItemIndex < 0) {
            this.firstItemIndex = this.capacity - 1
        }
        this.array[this.firstItemIndex] = item
        this._size++
    }

    override fun addLast(item: T) {
        this.growIfRequired()
        this.array[this.lastItemIndex] = item
        this.lastItemIndex++
        if (this.lastItemIndex == this.capacity) {
            this.lastItemIndex = 0
        }
        this._size++
    }

    override fun addAt(index: Int, item: T) {
        this.growIfRequired()
        when (index) {
            0 -> addFirst(item)
            this._size -> addLast(item)
            else -> {
                this.assertIndex(index)
                this.lastItemIndex++
                if (this.lastItemIndex == this.capacity) {
                    this.lastItemIndex = 0
                }
                this._size++
                for (remapIndex in this._size - 1 downTo index + 1) {
                    this[remapIndex] = this[remapIndex - 1]
                }
                this[index] = item
            }
        }
    }

    private fun assertNotEmpty() {
        if (this.isEmpty) {
            throw IllegalStateException("dequeue is empty")
        }
    }

    override fun removeFirst(): T {
        assertNotEmpty()
        val item = this[0]
        this.firstItemIndex++
        if (this.firstItemIndex == this.capacity) {
            this.firstItemIndex = 0
        }
        this._size--
        this.shrinkIfRequired()
        return item
    }

    override fun removeLast(): T {
        assertNotEmpty()
        val item = this[this._size - 1]
        this.lastItemIndex--
        if (this.lastItemIndex < 0) {
            this.lastItemIndex = this.capacity - 1
        }
        this._size--
        this.shrinkIfRequired()
        return item
    }

    override fun removeAt(index: Int): T {
        this.assertNotEmpty()
        when (index) {
            0 -> return this.removeFirst()
            this._size - 1 -> return this.removeLast()
            else -> {
                for (swapIndex in index until this._size - 1) {
                    this.swap(swapIndex, swapIndex + 1)
                }
                return this.removeLast()
            }
        }
    }

    private inner class ArrayDequeueIterator : Iterator<T> {
        private var index = 0

        override fun hasNext(): Boolean {
            return this.index < this@ArrayDeque._size
        }

        override fun next(): T {
            return this@ArrayDeque[this.index++]
        }
    }

    override fun iterator(): Iterator<T> {
        return ArrayDequeueIterator()
    }
}

File: src/algorithms/alg01deque/array_deque.py
from typing import TypeVar, Union, List, cast, Iterable

from algorithms.alg01deque.deque import Deque

T = TypeVar("T")


class Empty:
    pass


EMPTY = Empty()


class ArrayDeque(Deque[T]):
    def __init__(self, iterable: Iterable[T] = None, capacity: int = 32):
        super().__init__()
        self._array: List[Union[Empty, T]] = [EMPTY] * capacity
        self._head_item_index: int = 0
        self._tail_item_index: int = 0
        if iterable:
            for item in iterable:
                self.push_tail(item)

    @property
    def capacity(self) -> int:
        return len(self._array)

    def _assert_index(self, index: int):
        if self.is_empty:
            raise ValueError(f"cannot get index {index}: deque is empty")
        if index < 0 or index >= self.size:
            raise ValueError(f"want index in [0, {self.size - 1}]; have index={index}")

    def _normalize_index(self, index: int) -> int:
        self._assert_index(index)
        return self._head_item_index + index - self.capacity

    def __getitem__(self, index: int) -> T:
        return cast(T, self._array[self._normalize_index(index)])

    def __setitem__(self, index: int, item: T) -> None:
        self._array[self._normalize_index(index)] = item

    def _resize(self, capacity: int):
        array: List[Union[Empty, T]] = [EMPTY] * capacity
        for index in range(self.size):
            array[index] = self[index]
        self._array = array
        self._head_item_index = 0
        self._tail_item_index = self.size

    def _grow_if_required(self) -> None:
        if self.size == self.capacity:
            self._resize(self.capacity * 2)

    def _shrink_if_required(self) -> None:
        if self.size > 0 and self.size == self.capacity // 4:
            self._resize(self.capacity // 2)

    def push_head(self, item: T) -> None:
        self._grow_if_required()
        self._head_item_index -= 1
        if self._head_item_index < 0:
            self._head_item_index = self.capacity - 1
        self._array[self._head_item_index] = item
        self._size += 1

    def push_tail(self, item: T) -> None:
        self._grow_if_required()
        self._array[self._tail_item_index] = item
        self._tail_item_index += 1
        if self._tail_item_index == self.capacity:
            self._tail_item_index = 0
        self._size += 1

    def push_at(self, index: int, item: T) -> None:
        self._grow_if_required()
        if index == 0:
            self.push_head(item)
        elif index == self.size:
            self.push_tail(item)
        else:
            self._assert_index(index)
            self._tail_item_index += 1
            if self._tail_item_index == self.capacity:
                self._tail_item_index = 0
            self._size += 1
            for remap_index in range(self.size - 1, index, -1):
                self[remap_index] = self[remap_index - 1]
            self[index] = item

    def _assert_not_empty(self) -> None:
        if self.is_empty:
            raise AssertionError("dequeue is empty: cannot remove elements from it")

    def pop_head(self) -> T:
        self._assert_not_empty()
        item: T = self[0]
        self._head_item_index += 1
        if self._head_item_index == self.capacity:
            self._head_item_index = 0
        self._size -= 1
        self._shrink_if_required()
        return item

    def pop_tail(self) -> T:
        self._assert_not_empty()
        item: T = self[self.size - 1]
        self._tail_item_index -= 1
        if self._tail_item_index < 0:
            self._tail_item_index = self.capacity - 1
        self._size -= 1
        self._shrink_if_required()
        return item

    def pop_at(self, index: int) -> T:
        if index == 0:
            return self.pop_head()
        elif index == self.size - 1:
            return self.pop_tail()
        else:
            for swap_index in range(index, self.size - 1):
                self.swap(swap_index, swap_index + 1)
            return self.pop_tail()
